WalletWasabi/Crypto/Strobe.cs

```diff
@@ -0,0 +1,318 @@
+		{
+			BeginOperation(StrobeFlags.M | StrobeFlags.A, more);
+			Absorb(data);
+		}
+
+		public void AssociatedData(byte[] data, bool more)
+		{
+			BeginOperation(StrobeFlags.A, more);
+			Absorb(data);
+		}
+
+		public byte[] Prf(int count, bool more)
+		{
+			BeginOperation(StrobeFlags.I | StrobeFlags.A | StrobeFlags.C, more);
+			return Squeeze(new byte[count]);
+		}
+
+		public void Key(byte[] data, bool more)
+		{
+			BeginOperation(StrobeFlags.A | StrobeFlags.C, more);
+			Override(data);
+		}
+
+		public override string ToString()
+		{
+			return ByteHelpers.ToHex(_state);
+		}
+
+		private void Absorb(byte[] data)
+		{
+			foreach (var b in data)
+			{
+				_state[_position++] ^= b;
+				if (_position == SpongeRate)
+				{
+					RunF();
+				}
+			}
+		}
+
+		private void Override(byte[] data)
+		{
+			foreach (var b in data)
+			{
+				_state[_position++] = b;
+				if (_position == SpongeRate)
+				{
+					RunF();
+				}
+			}
+		} 
+
+		private byte[] Squeeze(byte[] data)
+		{
+			for (var i = 0; i < data.Length; i++)
+			{
+				data[i] = _state[_position];
+				_state[_position++] = 0;
+				if (_position == SpongeRate)
+				{
+					RunF();
+				}
+			}
+			return data;
+		}
+
+		private void BeginOperation(StrobeFlags flags, bool more)
+		{
+			if (more)
+			{
+				if (flags != _currentFlags)
+				{
+					throw new InvalidOperationException($"Attempt to continue operation '{_currentFlags}' with new flags '{flags}' is not allowed.");
+				}
+				_currentFlags = flags;
+			}
+
+			// Skip adjusting direction information (we just use AD, PRF)
+			if (_currentFlags.HasFlag(StrobeFlags.T))
+			{
+				throw new NotImplementedException("Transport operations are not implemented.");
+			}
+
+			var oldBeginPosition = _beginPosition;
+			_currentFlags = flags;
+			_beginPosition = (byte)(_position + 1);
+
+			Absorb(new[]{ oldBeginPosition, (byte)flags });
+
+			// Force running F if C or K is set
+			var forceF = flags.HasFlag(StrobeFlags.C) || flags.HasFlag(StrobeFlags.K);
+
+			if (forceF && _position != 0)
+			{
+				RunF();
+			}
+		}
+
+		private void RunF()
+		{
+			_state[_position] ^= _beginPosition;
+			_state[_position + 1] ^= DDATA;
+			_state[SpongeRate + 1] ^= DRATE;
+			
+			KeccakF1600(_state);
+
+			_position = 0;
+			_beginPosition = 0;
+		}
+
+
+		private static readonly ulong[] KeccakRoundConstants = new ulong[]{
+			0x0000000000000001UL, 0x0000000000008082UL, 0x800000000000808aUL, 0x8000000080008000UL,
+			0x000000000000808bUL, 0x0000000080000001UL, 0x8000000080008081UL, 0x8000000000008009UL,
+			0x000000000000008aUL, 0x0000000000000088UL, 0x0000000080008009UL, 0x000000008000000aUL,
+			0x000000008000808bUL, 0x800000000000008bUL, 0x8000000000008089UL, 0x8000000000008003UL,
+			0x8000000000008002UL, 0x8000000000000080UL, 0x000000000000800aUL, 0x800000008000000aUL,
+			0x8000000080008081UL, 0x8000000000008080UL, 0x0000000080000001UL, 0x8000000080008008UL
+		};
+
+		static private void KeccakF1600(byte[] state)
```

Comment:
If this is from Bouncy Castle as I suspect it should probably contain copyright and license information (maybe best to put it in a separate file?) and a comment with a permalink to the version that was copied, so that would allow people to compare the implementations

Reply:
I will not put it in a separate file because we will not use it anywhere else and because it is not a class, in fact I extracted and adapted only this method.